### DISCUSSION TITLE
replace welcome page by workspace readme at installation time

### DISCUSSION THREAD

**Discussion by hecnavsanz** [coder/code-server#4567] [Q&A] [answered]
After installing code-server, I've created a workspace that it's mapping to a directory with a _readme.md_ file and setup in the _$HOME/.local/share/code-server/User/Workspaces/ws-name.code-workspace_ file this entry:

`"settings": {
                "workbench.startupEditor": "readme"
        }`

and in the _$HOME/.local/share/code-server/User/settings.json_ this other entry:

`"workbench.startupEditor": "none"`

When I'm accessing the first time to code-server I'm getting the Welcome page .. not my _readme.md_ page. I'm trying but no way to setup code-server to display my _readme.md_ page and not the Welcome page.

Thnx in adv.

**Comment by hecnavsanz** [comment]
Also just for curiosity, at FAQs suggest to set `workspace `in the `query `entry in the _coder.json_ file ... but even doing so ... when accessing the first time the entry is removed. My _coder.json_ : 

`{
  "query": {
    "workspace": "$HOME/.local/share/code-server/User/Workspaces/ws-name.code-workspace"
  },
  "lastVisited": {
    "url": "$HOME/.local/share/code-server/User/Workspaces/ws-name.code-workspace",
    "workspace": true
  },
  "update": {
    "checked": 1638291450345,
    "version": "3.12.0"
  }
}`

Thnx again.

**Comment by jsjoeio** [comment]
Hmm...what version of code-server are you using? Can you post basic repro steps? Happy to try myself locally. I don't use Workspaces though so repro steps would help!

**Reply by hecnavsanz** [reply]
v.3.12

steps: 

1. Install code-server.
2. copy custom _config.yaml_ to _/home/user/.config/code-server_ directory:
`bind-addr: host1:8080`
`auth: none`
`cert: false`
3. Start code-server : _systemctl start code-server@user_
4. Install VSCode extensions in code-server:
`code-server --install-extension ms-python.python`
`code-server --install-extension kelvin.vscode-sshfs`
5. Setup codr, user and workspace settings:
**codr**
`mkdir /home/user/.config/code-server`
`cp coder.json /home/user/.config/code-server`
with _coder.json_:
`{
  "query": {
    "folder": "/home/user/.local/share/code-server/User/Workspaces/ws-name.code-workspace"
  },
  "lastVisited": {
    "url": "/home/user/.local/share/code-server/User/Workspaces/ws-name.code-workspace",
    "workspace": true
  }
}`
and ...
**user**
`mkdir /home/user/.local/share/code-server/User`
`cp settings.json /home/user/.local/share/code-server/User`
with _settings.json_:
`{
    "workbench.colorTheme": "Default Dark+",
    "workbench.startupEditor": "readme",
    "files.exclude": {
        "**/.*": true
    }`
and ...
**workspace**
`mkdir /home/user/.local/share/code-server/User/Workspaces`
`cp settings.code-workspace /home/user/.local/share/code-server/User/Workspaces/ws-name.code-workspace`
with _settings.code-workspace_:
`{
        "folders": [
                {
                        "name": "host1",
                        "path": "/home/user"
                }
        ]
}`
6. use a sample readme.md file in /home/user.

Maybe it's more simple than this ... basically I want to after the installation open a custom _readme.md_ file located in $HOME or other directory when first time access to codr ... instead of the end-user to see the Welcome page and enter into guided steps.

I'm creating Vagrant boxed and automating w/ Ansible the installation process ... to create a lab env w/ multiple machines so students can use a controller vm w/ everything to access remotely to other machines and a _readme.md_ with the lab instructions.

Thanks in adv.

**Reply by jsjoeio** [reply]
Hmm...I feel like we can simplify this a lot more:
1. install code-server
2. run code-server with `settings.json` suggested
3. open workspace which has `readme`

Expected: see `readme`
Actual: see Welcome page

With that in mind, let's test!

## Video 

https://user-images.githubusercontent.com/3806031/144891899-d1aafc24-6f84-4b55-88a3-bfbde60eddd7.mov

I even made this test dir a workspace and it works as expected.

![image](https://user-images.githubusercontent.com/3806031/144891964-5e6da821-3a54-4d30-af9a-ddab2d8b4f07.png)

> a readme.md with the lab instructions

As long as you have the right `settings.json` and a `readme` in the lab repo, then it _should_ work as you want.

**Comment by hecnavsanz** [comment, marked as answer]
Thnx Joe .. it gave to me some ideas to solve this issue:

Basically to automate this when deploying code-server:
1. Install code-server: `curl -fsSL https://code-server.dev/install.sh | sh`
2. Install code-server extensions:  `code-server --install-extension ...`
3. Deploy custom `config.yaml` at `/home/<user>/.config/code-server` directory:
```
--- config.yaml
bind-addr: <host_pub_ip>:443
auth: none
cert: true
```
4. Create workspaces directory: `mkdir /home/<user>/.local/share/code-server/User/Workspaces`
5. Deploy custom `coder.json` at `/home/<user>/.local/share/code-server`:
```
# coder.json
{
  "query": {
    "folder": "/labs"
  },
  "lastVisited": {
    "url": "/home/<user>/.local/share/code-server/User/Workspaces/labs.code-workspace",
    "workspace": true
  }
}
```
  Note: the `/labs` directory contains the `readme.md` markdown file that will preview automatically.
6. Deploy custom `settings.json` at `/home/<user>/.local/share/code-server/User`:
```
# settings.json
{
        "workbench.startupEditor": "readme"
}
```
7. Deploy custom `labs.code-workspace` at `/home/<user>/.local/share/code-server/User/Workspaces`:
```
# settings.code-workspace
{
        "folders": [
                {
                        "name": "<name>",
                        "path": "/labs"
                }
        ]
}
```
8. Start code-server: `systemctl start code-server@<user>`
9. Verify:
![coder](https://user-images.githubusercontent.com/82879431/146418938-c824c65f-d727-43f5-8470-05727b0d2e1c.png)

Thanks again.

**Reply by jsjoeio** [reply]
Oh wow! This is an awesome workflow. Thanks for taking the time to write it up and share with us!

cc @bpmct I think you'll find this interesting.
